server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java

```diff
@@ -472,10 +474,79 @@ private void executeSearch(SearchTask task, SearchTimeProvider timeProvider, Sea
         Map<String, Set<String>> routingMap = indexNameExpressionResolver.resolveSearchRouting(clusterState, searchRequest.routing(),
             searchRequest.indices());
         routingMap = routingMap == null ? Collections.emptyMap() : Collections.unmodifiableMap(routingMap);
-        String[] concreteIndices = new String[indices.length];
-        for (int i = 0; i < indices.length; i++) {
-            concreteIndices[i] = indices[i].getName();
+        Map<String, Float> concreteIndexBoosts = resolveIndexBoosts(searchRequest, clusterState);
+
+        if (searchRequest.scroll() != null || searchRequest.searchType() == DFS_QUERY_THEN_FETCH
+            || searchRequest.indicesOptions().ignoreThrottled()) {
+            String[] concreteIndices = Arrays.stream(indices).map(Index::getName).toArray(String[]::new);
+            executeSearch(task, timeProvider, searchRequest, localIndices, concreteIndices, routingMap,
+                aliasFilter, concreteIndexBoosts, remoteShardIterators, remoteConnections, clusterState, listener, clusters);
+        } else {
+            List<String> throttledIndices = new ArrayList<>();
+            List<String> nonThrottledIndices = new ArrayList<>();
+            for (Index index : indices) {
+                IndexMetaData indexMetaData = clusterState.getMetaData().index(index);
+                if (IndexSettings.INDEX_SEARCH_THROTTLED.get(indexMetaData.getSettings())) {
+                    throttledIndices.add(index.getName());
+                } else {
+                    nonThrottledIndices.add(index.getName());
+                }
+            }
```

Comment:
here we could do things differently, here are a couple of ideas:
1) distinguish between read-only indices and indices that are being written to (suggested by @jimczi ). That may complicate setting pre_filter_shard_size to `1` only for frozen indices though.
2) this PR considers remote indices as fast as local ones, but maybe they should be treated as slow ones like the throttled ones?

Reply:
I think it's a good default to start with throttling vs. non-throttled. You could go and check index blocks instead and if it's read-only we group them into a third group? I do wonder if it makes sense to simply have a list of groups rather than a separation based on throttled and non-throttled only.

Reply:
on the groups suggestion: some of the non-throttled indices may fall into this third read-only group, right? Do we want to run three searches in that case then? one for throttled, one for non-throttled read-only, and one for non-throttled non-read-only?

Reply:
Read-only seem to fall into the throttled group IMO.

Reply:
as I said above, one thing about having throttled indices only in one group is that we can set pre_filter_shard_size to 1 automatically for those indices, which is much better than telling users they should set it to 1 anytime they are searching against throttled indices. But setting it to 1 should not be done for all read-only indices. Shall we have three groups: throttled, read-only non-throttled and non-throttled? Or two groups throttled and non-throttled? Or ignore the pre_filter_shard_size problem for now?

Reply:
I went for not addressing the pre_filter_shard_size issue for throttled indices. We will discuss that separately.